Add the `fermi` module docs (#4250)
PennyLaneAI/pennylane: doc/code/qml_fermi.rst

qml.fermi
=========

Overview
--------

This module contains functions and classes for creating and manipulating fermionic operators.

.. currentmodule:: pennylane.fermi

.. automodapi:: pennylane.fermi
    :no-heading:
    :no-main-docstr:

>     :no-inherited-members:

Reviewer: Is the plan to have `qml.jordan_wigner` at top level? I couldn't find it in the built docs.

and :class:`~pennylane.FermiA`, respectively, by passing the index of the orbital that the fermionic
operator acts on. For instance, the operators :math:`a^{\dagger}_0` :math:`a_3` are constructed as

>>> qml.FermiC(0)
>>> qml.FermiA(3)

These operators can be multiplied by each other to create :math:`n`-orbital fermionic operators such
as :math:`a^{\dagger}_0 a_0 a^{\dagger}_3 a_3` that we call a Fermi word.

>>> qml.FermiC(0) * qml.FermiA(0) * qml.FermiC(3) * qml.FermiA(3)
<FermiWord = '0+ 0- 3+ 3-'>

The Fermi words can be linearly combined to create fermionic operators that we call a Fermi
Sentence. For instance, a fermionic Hamiltonian such as
:math:`H = 1.2 a^{\dagger}_0 a_0  + 2.3 a^{\dagger}_3 a_3` can be constructed as

>>> h = 1.2 * qml.FermiC(0) * qml.FermiA(0) + 2.3 * qml.FermiC(3) * qml.FermiA(3)
>>> h
1.2 * '0+ 0-'
+ 2.3 * '3+ 3-'

These fermionic objects can be mapped to the qubit basis by using the function
:func:`jordan_wigner`

>>> qml.jordan_wigner(h)
((-1.75+0j)*(Identity(wires=[0]))) + ((0.6+0j)*(PauliZ(wires=[0]))) + ((1.15+0j)*(PauliZ(wires=[3])))

FermiWord and FermiSentence
---------------------------

> Fermi words and Fermi Sentences can also be constructed directly with

Reviewer 2: ```suggestion
Fermi words and Fermi sentences can also be constructed directly with
```

Is there a reason to capitalize Sentence but not word?

:class:`~pennylane.fermi.FermiWord` and :class:`~pennylane.fermi.FermiSentence` by passing
dictionaries that define the fermionic operators.

> For Fermi words, the dictionary items identify the fermionic creation and annihilation operators.

Reviewer 2: ```suggestion
For Fermi words, the dictionary items define fermionic creation and annihilation operators.
```

I'm not sure this is the best way to phrase it either, but I was confused by 'identify'

> Each operators is identified by a tuple of two integers representing its position in the Fermi word

Reviewer 2: ```suggestion
Each annihilation/creation operator is defined by a tuple of two integers representing its position in the Fermi word and
```

the orbital it acts on, and a symbol identifying its type. We use ``'+'`` and ``'-'`` to denote
creation and annihilation operators, respectively. The operator
:math:`a^{\dagger}_0 a_0 a^{\dagger}_3 a_3` can then be constructed with

> >>> qml.fermi.FermiWord({(0, 0): '+', (1, 3): '-'})

Reviewer 2: Could we make it `>>> qml.fermi.FermiWord({(0, 0): '+', (1, 3): '-', (2, 1): '+'})`? I think having the third entry more clearly solidifies the 0, 1, 2... pattern for the first entry in the tuple

<FermiWord = '0+ 3-'>

A Fermi sentence can be constructed directly by passing a dictionary of Fermi words and their
corresponding coefficients to the :class:`~pennylane.fermi.FermiSentence` class. For instance, the
Fermi sentence :math:`1.2 a^{\dagger}_0 a_0  + 2.3 a^{\dagger}_3 a_3` can be constructed as

>>> fw1 = qml.fermi.FermiWord({(0, 0): '+', (1, 0): '-'})
>>> fw2 = qml.fermi.FermiWord({(0, 3): '+', (1, 3): '-'})
>>> qml.FermiSentence({fw1: 1.2, fw2, 2.3})
1.2 * '0+ 0-'
+ 2.3 * '3+ 3-'

Mapping to qubit operators
--------------------------

The fermionic operators can be mapped to the qubit basis by using the

> :func:`~pennylane.jordan_wigner` function. This function can be used to map creation and

Reviewer 2: > This function can be used to map creation and annihilation operators as well as Fermi words and Fermi Sentences.

This sentence confused me a bit. I would be inclined to either drop it (because I think its clear in the examples below) or use the class names (`FermiA`, `FermiC`, `FermiWord`, `FermiSentence` instead)

Author: I moved that section above, so I cannot use `FermiWord` and  `FermiSentence` here. Is it still confusing after the refactoring?

Reviewer 2: No, I think when it comes immediately after the explanation for creation and annihilation operators and Fermi words and Fermi sentences, its quite clear :)

annihilation operators as well as Fermi words and Fermi Sentences.

>>> qml.jordan_wigner(qml.FermiA(1))
(0.5*(PauliZ(wires=[0]) @ PauliX(wires=[1]))) + (0.5j*(PauliZ(wires=[0]) @ PauliY(wires=[1])))

>>> qml.jordan_wigner(qml.FermiC(1) * qml.FermiA(1))
((0.5+0j)*(Identity(wires=[1]))) + ((-0.5+0j)*(PauliZ(wires=[1])))

>>> qml.jordan_wigner(0.5 * qml.FermiC(1) * qml.FermiA(1) + 0.75 * qml.FermiC(2) * qml.FermiA(2))

> ((0.625+0j)*(Identity(wires=[1]))) + ((-0.25+0j)*(PauliZ(wires=[1]))) + ((-0.375+0j)*(PauliZ(wires=[2])))

Author: With this section brought up before "FermiWord and FermiSentence" section, we cannot mention this here. I think the docstring of the function is clear enough on that, probably no need to mention this info here?

> ((0.625+0j)*(Identity(wires=[1]))) + ((-0.25+0j)*(PauliZ(wires=[1]))) + ((-0.375+0j)*(PauliZ(wires=[2])))

Reviewer 2: Why not? `PauliSentence` is a separate thing that doesn't require familiarity with `FermiWord` and `FermiSentence`. 

But I agree the docstring is clear, and the default behaviour is probably what most people want, so I'm happy to leave it out.

qml.fermi
=========

Overview
--------

This module contains functions and classes for creating and manipulating fermionic operators.

.. currentmodule:: pennylane.fermi

.. automodapi:: pennylane.fermi
    :no-heading:
    :no-main-docstr:
    :no-inherited-members:

FermiC and FermiA
-----------------

> The fermionic creation and annihilation operators are constructed with :class:`~pennylane.FermiC`
> and :class:`~pennylane.FermiA`, respectively, by passing the index of the orbital that the fermionic
> operator acts on. For instance, the operators :math:`a^{\dagger}_0` and :math:`a_3` are constructed as

Reviewer 3: I think this could be easier to follow. Maybe something like:
```suggestion
The fermionic creation and annihilation operators are constructed with :class:`~pennylane.FermiC`
and :class:`~pennylane.FermiA`. We pass in the index of the orbital that the
operator acts on. For example, the operators :math:`a^{\dagger}_0` and :math:`a_3`, acting on the 0th and 3rd orbitals, are constructed as
```

>>> qml.FermiC(0)
a⁺(0)
>>> qml.FermiA(3)
a(3)

> These operators can be multiplied by each other to create :math:`n`-orbital fermionic operators such
> as :math:`a^{\dagger}_0 a_0 a^{\dagger}_3 a_3` that we call a Fermi word.

Reviewer 3: ```suggestion
These operators can be multiplied by each other to create :math:`n`-orbital fermionic operators such
as :math:`a^{\dagger}_0 a_0 a^{\dagger}_3 a_3`. We call these :math:`n`-orbital fermionic operators "Fermi words".
```

>>> qml.FermiC(0) * qml.FermiA(0) * qml.FermiC(3) * qml.FermiA(3)
a⁺(0) a(0) a⁺(3) a(3)

The Fermi words can be linearly combined to create fermionic operators that we call a Fermi

> Sentence. For instance, a fermionic Hamiltonian such as

Reviewer 2: ```suggestion
sentence. For instance, a fermionic Hamiltonian such as
```
Looks like we're going with lowercase everywhere else, but this one is still capitalized

> The Fermi words can be linearly combined to create fermionic operators that we call a Fermi
> sentence. For instance, a fermionic Hamiltonian such as
> :math:`H = 1.2 a^{\dagger}_0 a_0  + 2.3 a^{\dagger}_3 a_3` can be constructed as

Reviewer 3: ```suggestion
The Fermi words can be linearly combined to create fermionic operators that we call a Fermi
sentence. For instance, a fermionic Hamiltonian such as
:math:`H = 1.2 a^{\dagger}_0 a_0  + 2.3 a^{\dagger}_3 a_3` can be constructed with a Fermi word as follows:
```

>>> h = 1.2 * qml.FermiC(0) * qml.FermiA(0) + 2.3 * qml.FermiC(3) * qml.FermiA(3)
>>> h
1.2 * a⁺(0) a(0)
+ 2.3 * a⁺(3) a(3)

Mapping to qubit operators
--------------------------

> The fermionic operators can be mapped to the qubit basis by using the
> :func:`~pennylane.jordan_wigner` function. This function can be used to map creation and
> annihilation operators as well as Fermi words and Fermi sentences.

Reviewer 3: ```suggestion
The fermionic operators can be mapped to the qubit basis by using the
:func:`~pennylane.jordan_wigner` function. This function can be used to map :class:`~pennylane.FermiC` and
:class:`~pennylane.FermiA` operators as well as Fermi words and Fermi sentences.
```

>>> qml.jordan_wigner(qml.FermiA(1))
(0.5*(PauliZ(wires=[0]) @ PauliX(wires=[1])))
+ (0.5j*(PauliZ(wires=[0]) @ PauliY(wires=[1])))

>>> qml.jordan_wigner(qml.FermiC(1) * qml.FermiA(1))
((0.5+0j)*(Identity(wires=[1])))
+ ((-0.5+0j)*(PauliZ(wires=[1])))

>>> f = 0.5 * qml.FermiC(1) * qml.FermiA(1) + 0.75 * qml.FermiC(2) * qml.FermiA(2)
>>> qml.jordan_wigner(f)
((0.625+0j)*(Identity(wires=[1])))
+ ((-0.25+0j)*(PauliZ(wires=[1])))
+ ((-0.375+0j)*(PauliZ(wires=[2])))

FermiWord and FermiSentence
---------------------------

Fermi words and Fermi sentences can also be constructed directly with
:class:`~pennylane.fermi.FermiWord` and :class:`~pennylane.fermi.FermiSentence` by passing
dictionaries that define the fermionic operators.

For Fermi words, the dictionary items define the fermionic creation and annihilation operators.
Each creation or annihilation operator is defined by a tuple of two integers representing its
position in the Fermi word and the orbital it acts on, and a symbol identifying its type. We use
``'+'`` and ``'-'`` to denote creation and annihilation operators, respectively. The operator

> :math:`a^{\dagger}_0 a_0 a^{\dagger}_3 a_3` can then be constructed with

Reviewer 3: So the keys of the dictionary are the tuples and the values are a tag determining whether the operator is an annihilation or creator operator. This could be expressed more directly here.

Author: Rephrased.

qml.fermi
=========

Overview
--------

This module contains functions and classes for creating and manipulating fermionic operators.

.. currentmodule:: pennylane.fermi

.. automodapi:: pennylane.fermi
    :no-heading:
    :no-main-docstr:

>     :no-inherited-members:

Reviewer: Could we have it so that there is not multiple instances of the same function in the doc?
E.g., [this](https://xanaduai-pennylane--4250.com.readthedocs.build/en/4250/code/api/pennylane.FermiA.html) and [this](https://xanaduai-pennylane--4250.com.readthedocs.build/en/4250/code/api/pennylane.fermi.FermiA.html)?

To do this, we need to change here so that `qml.fermi.FermiA` is not included in the docs and we instead point explicitly to `qml.FermiA`.

Author: Yes, will address in a new PR.

FermiC and FermiA
-----------------

The fermionic creation and annihilation operators are constructed with :class:`~pennylane.FermiC`

> and :class:`~pennylane.FermiA`. We pass in the index of the orbital that the

Reviewer 4: ```suggestion
and :class:`~pennylane.FermiA`, respectively. We pass in the index of the orbital that the
```

operator acts on. For example, the operators :math:`a^{\dagger}_0` and :math:`a_3`, acting on the

> :math:`0`th and :math:`3`rd orbitals, are constructed as

Reviewer 2: ```suggestion
:math:`0`\th and :math:`3`\rd orbitals, are constructed as
```

I _think_ this will fix the failing documentation thing. Otherwise:

```suggestion
:math:`0\text{th}` and :math:`3\text{rd}` orbitals, are constructed as
```
